api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -400,6 +400,12 @@ Builder checkConflicts(boolean check) {
     private void checkAndAddPartitionName(String name, Integer sourceColumnId) {
       Types.NestedField schemaField =
           this.caseSensitive ? schema.findField(name) : schema.caseInsensitiveFindField(name);
+      // corner case: extra partition column in metadata table and whose field id is extremely large
+      if (null != schemaField
+          && schemaField.name().equals("partition")
+          && schemaField.fieldId() > 2000) {
```

Comment:
This change feels brittle: it hard-codes a special case for a field named `partition` and uses a heuristic (>2000) to infer metadata field ids.

Reply:
Yeah I'd need to double check what's exactly breaking but all this resolution should be based on ids anyways and the partition metadata field ID would be different